File: harvest.py
#!/usr/bin/env python3
"""TASK LORD harvester — clock every project's loose ends into one Kanban board.

Pipeline:
  1. Collect raw records from three sources (sessions, code, Trello).
  2. Merge git 'cold facts' onto session records sharing a cwd.
  3. For each session/code record, ask local Ollama to classify status and
     reconstruct where it left off (the token-grind — never burns Claude).
  4. Emit board.json the single-file HTML board reads.

Run:  python3 harvest.py            (full run)
      python3 harvest.py --no-llm   (heuristics only, fast, no Ollama)
"""
from __future__ import annotations

import argparse
import json
import sys
from datetime import datetime, timezone
from pathlib import Path

import ollama_client
import sources_code
import sources_trello
import sources_transcripts
import store

HERE = Path(__file__).resolve().parent
BOARD_JSON = HERE / "board" / "board.json"
VALID_STATUS = {"backlog", "in_progress", "blocked", "done"}

SUMMARY_PROMPT = """You are a project triage assistant for a founder's software factory.
Given the tail of a work session and git facts for ONE project, output STRICT JSON:
{{
  "status": one of "backlog" | "in_progress" | "blocked" | "done",
  "left_off": "1-2 sentences: concretely where this was left off / last state",
  "next_step": "the single most useful next action, imperative voice",
  "blockers": "what's blocking it, or empty string if nothing"
}}

Rules:
- "done" only if the work clearly reached a finished, shipped, or verified state.
- "blocked" if it's waiting on a person, decision, credential, or external thing.
- "in_progress" if mid-build with a clear unfinished thread.
- "backlog" if barely started or just scoped.
- Be specific and concrete. No preamble. JSON only.

PROJECT: {label}
GIT FACTS: {facts}
SESSION TAIL:
{tail}
"""


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _days_since(iso: str | None) -> int | None:
    if not iso:
        return None
    try:
        dt = datetime.fromisoformat(iso.replace("Z", "+00:00"))
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return (datetime.now(timezone.utc) - dt).days
    except (ValueError, TypeError):
        return None


def _facts_str(rec: dict) -> str:
    bits = []
    if rec.get("is_git"):
        bits.append(f"git branch={rec.get('branch')}")
        if rec.get("last_commit_msg"):
            bits.append(f"last commit: {rec['last_commit_msg']!r}")
        if rec.get("dirty"):
            bits.append(f"{rec.get('dirty_count', 0)} uncommitted changes")
    if rec.get("plan_files"):
        bits.append("has " + ", ".join(rec["plan_files"]))
    age = _days_since(rec.get("last_activity"))
    if age is not None:
        bits.append(f"last active {age}d ago")
    return "; ".join(bits) or "none"


def _heuristic(rec: dict) -> dict:
    """Status guess without an LLM — fallback when Ollama is down."""
    age = _days_since(rec.get("last_activity"))
    tail = rec.get("tail", "")
    low = tail.lower()
    if any(w in low for w in ("blocked", "waiting on", "need eric", "can't proceed")):
        status = "blocked"
    elif any(w in low for w in ("shipped", "done.", "complete", "deployed", "verified")):
        status = "done"
    elif age is not None and age > 45:
        status = "backlog"
    else:
        status = "in_progress"
    last_user = ""
    for chunk in reversed(tail.split("\n\n")):
        if chunk.startswith("USER:"):
            last_user = chunk[5:].strip()[:200]
            break
    return {
        "status": status,
        "left_off": last_user or "No recent summary available.",
        "next_step": "Review session and resume.",
        "blockers": "" if status != "blocked" else "See session tail.",
    }


def _summarize(rec: dict, use_llm: bool) -> dict:
    if not use_llm:
        return _heuristic(rec)
    prompt = SUMMARY_PROMPT.format(
        label=rec.get("label", "?"),
        facts=_facts_str(rec),
        tail=rec.get("tail", "(no session tail)"),
    )
    result = ollama_client.generate_json(prompt)
    if not result or result.get("status") not in VALID_STATUS:
        return _heuristic(rec)
    return {
        "status": result.get("status", "in_progress"),
        "left_off": (result.get("left_off") or "").strip()[:400] or "—",
        "next_step": (result.get("next_step") or "").strip()[:300] or "—",
        "blockers": (result.get("blockers") or "").strip()[:300],
    }


def build(use_llm: bool = True) -> dict:
    sessions = sources_transcripts.collect()
    code = sources_code.collect()
    print(f"[harvest] {len(sessions)} session groups, {len(code)} code projects", file=sys.stderr)

    # Merge git facts onto sessions by path; track which code paths are covered.
    covered = set()
    for rec in sessions:
        path = rec.get("path")
        if path and path in code:
            rec.update({k: v for k, v in code[path].items() if k != "label"})
            covered.add(path)

    # Code projects with NO recent session → cold cards (still loose ends).
    for path, facts in code.items():
        if path in covered:
            continue
        sessions.append(
            {
                "source": "code",
                "label": facts["label"],
                "path": path,
                "last_activity": facts.get("last_commit"),
                "tail": "",
                **{k: v for k, v in facts.items() if k != "label"},
            }
        )

    cards = []
    total = len(sessions)
    for i, rec in enumerate(sessions, 1):
        summ = _summarize(rec, use_llm=use_llm and bool(rec.get("tail")))
        if not rec.get("tail") and not use_llm:
            summ = _heuristic(rec)
        age = _days_since(rec.get("last_activity"))
        cards.append(
            {
                "id": rec.get("group") or rec.get("path") or rec.get("label"),
                "label": rec.get("label", "?"),
                "source": rec.get("source", "session"),
                "status": summ["status"],
                "left_off": summ["left_off"],
                "next_step": summ["next_step"],
                "blockers": summ["blockers"],
                "path": rec.get("path"),
                "session_id": rec.get("session_id"),
                "last_activity": rec.get("last_activity"),
                "days_idle": age,
                "stale": age is not None and age > 30,
                "git": {
                    "branch": rec.get("branch"),
                    "dirty": rec.get("dirty"),
                    "dirty_count": rec.get("dirty_count"),
                    "last_commit_msg": rec.get("last_commit_msg"),
                }
                if rec.get("is_git")
                else None,
                "msg_count": rec.get("msg_count"),
            }
        )
        if i % 10 == 0 or i == total:
            print(f"[harvest] summarized {i}/{total}", file=sys.stderr)

    # Trello business cards (already classified, no LLM needed).
    for t in sources_trello.collect():
        age = _days_since(t.get("last_activity"))
        cards.append(
            {
                "id": "trello:" + t["label"],
                "label": t["label"],
                "source": "business",
                "status": t.get("status", "backlog"),
                "left_off": t.get("left_off") or "—",
                "next_step": t.get("next_step") or "—",
                "blockers": "",
                "path": t.get("path"),
                "last_activity": t.get("last_activity"),
                "days_idle": age,
                "stale": age is not None and age > 30,
                "git": None,
                "trello_list": t.get("trello_list"),
                "board": t.get("board"),
                "due": t.get("due"),
            }
        )

    # Sort each card by recency (most recent first); None last.
    cards.sort(key=lambda c: c.get("last_activity") or "", reverse=True)

    counts = {s: sum(1 for c in cards if c["status"] == s) for s in VALID_STATUS}
    return {
        "generated_at": _now_iso(),
        "total": len(cards),
        "counts": counts,
        "cards": cards,
    }


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--no-llm", action="store_true", help="heuristics only, skip Ollama")
    args = ap.parse_args()

    use_llm = not args.no_llm
    if use_llm and not ollama_client.is_up():
        print("[harvest] Ollama not reachable — falling back to heuristics", file=sys.stderr)
        use_llm = False

    board = build(use_llm=use_llm)
    # SQLite is the source of truth (+ keeps status history); board.json/.js are
    # exported FROM the db for the static/file:// UI.
    summary = store.upsert_cards(board["cards"])
    print(f"[harvest] db upsert: {summary}", file=sys.stderr)
    _write_board(store.export_board())
    print(f"[harvest] wrote {store.DB_PATH.name} + {BOARD_JSON.name} — "
          f"{board['total']} cards {board['counts']}", file=sys.stderr)
    return 0


def _write_board(board: dict) -> None:
    """Write both board.json (served) and board.js (file:// double-click)."""
    BOARD_JSON.parent.mkdir(parents=True, exist_ok=True)
    payload = json.dumps(board, indent=2)
    BOARD_JSON.write_text(payload)
    # board.js lets index.html load via file:// without CORS-blocked fetch.
    (BOARD_JSON.parent / "board.js").write_text("window.BOARD = " + payload + ";\n")


if __name__ == "__main__":
    raise SystemExit(main())

File: src/harvest.rs
//! Orchestrator: sources -> summarize (local LLM) -> SQLite -> board.json.

use crate::model::{Card, Summary};
use crate::sources::{code, transcripts, trello};
use crate::{config, llm, store};
use anyhow::Result;
use chrono::{DateTime, Utc};
use std::fs;

const VALID: [&str; 4] = ["backlog", "in_progress", "blocked", "done"];

const SUMMARY_PROMPT: &str = r#"You are a project triage assistant for a founder's software factory.
Given the tail of a work session and git facts for ONE project, output STRICT JSON:
{"status": "backlog"|"in_progress"|"blocked"|"done",
 "left_off": "1-2 sentences: concretely where this was left off",
 "next_step": "the single most useful next action, imperative voice",
 "blockers": "what's blocking it, or empty string"}

Rules: "done" only if clearly finished/shipped/verified. "blocked" if waiting on a
person/decision/credential/external thing. "in_progress" if mid-build. "backlog" if
barely started. Be specific. JSON only.

PROJECT: {label}
GIT FACTS: {facts}
SESSION TAIL:
{tail}
"#;

fn days_since(iso: &Option<String>) -> Option<i64> {
    let s = iso.as_ref()?;
    let dt = DateTime::parse_from_rfc3339(s).ok()?;
    Some((Utc::now() - dt.with_timezone(&Utc)).num_days())
}

fn facts_str(c: &Card) -> String {
    let mut bits = Vec::new();
    if let Some(g) = &c.git {
        if let Some(b) = &g.branch {
            bits.push(format!("git branch={}", b));
        }
        if let Some(m) = &g.last_commit_msg {
            bits.push(format!("last commit: {:?}", m));
        }
        if g.dirty {
            bits.push(format!("{} uncommitted changes", g.dirty_count));
        }
    }
    if let Some(d) = c.days_idle {
        bits.push(format!("last active {}d ago", d));
    }
    if bits.is_empty() {
        "none".into()
    } else {
        bits.join("; ")
    }
}

fn heuristic(c: &Card) -> Summary {
    let low = c.tail.to_lowercase();
    let status = if ["blocked", "waiting on", "need eric", "can't proceed"].iter().any(|w| low.contains(w)) {
        "blocked"
    } else if ["shipped", "done.", "complete", "deployed", "verified"].iter().any(|w| low.contains(w)) {
        "done"
    } else if c.days_idle.map(|d| d > 45).unwrap_or(false) {
        "backlog"
    } else {
        "in_progress"
    };
    let chunks: Vec<&str> = c.tail.split("\n\n").collect();
    let last_user = chunks
        .iter()
        .rev()
        .find(|chunk| chunk.starts_with("USER:"))
        .map(|s| s.trim_start_matches("USER:").trim().chars().take(200).collect::<String>())
        .unwrap_or_default();
    Summary {
        status: status.into(),
        left_off: if last_user.is_empty() { "No recent summary available.".into() } else { last_user },
        next_step: "Review session and resume.".into(),
        blockers: if status == "blocked" { "See session tail.".into() } else { String::new() },
    }
}

async fn summarize(c: &Card, use_llm: bool) -> Summary {
    if use_llm && !c.tail.is_empty() {
        let prompt = SUMMARY_PROMPT
            .replace("{label}", &c.label)
            .replace("{facts}", &facts_str(c))
            .replace("{tail}", &c.tail);
        if let Some(s) = llm::generate_json::<Summary>(&prompt).await {
            if VALID.contains(&s.status.as_str()) {
                return Summary {
                    status: s.status,
                    left_off: trunc(&s.left_off, 400, "—"),
                    next_step: trunc(&s.next_step, 300, "—"),
                    blockers: s.blockers.trim().chars().take(300).collect(),
                };
            }
        }
    }
    heuristic(c)
}

fn trunc(s: &str, n: usize, dflt: &str) -> String {
    let t: String = s.trim().chars().take(n).collect();
    if t.is_empty() {
        dflt.into()
    } else {
        t
    }
}

pub async fn run(use_llm: bool) -> Result<()> {
    let mut sessions = transcripts::collect();
    let code = code::collect();
    eprintln!("[harvest] {} session groups, {} code projects", sessions.len(), code.len());

    // merge git facts onto sessions by path
    let mut covered = std::collections::HashSet::new();
    for c in sessions.iter_mut() {
        if let Some(p) = &c.path {
            if let Some(f) = code.get(p) {
                c.git = f.git.clone();
                covered.insert(p.clone());
            }
        }
    }
    // cold code projects (no recent session)
    for (path, f) in &code {
        if covered.contains(path) {
            continue;
        }
        sessions.push(Card {
            id: path.clone(),
            source: "code".into(),
            label: f.label.clone(),
            path: Some(path.clone()),
            git: f.git.clone(),
            last_activity: f.last_commit.clone(),
            ..Default::default()
        });
    }

    let use_llm = use_llm && llm::available().await;
    if use_llm {
        eprintln!("[harvest] summarizing via {}", llm::describe());
    } else {
        eprintln!("[harvest] LLM unavailable — using heuristics");
    }

    let total = sessions.len();
    let mut cards: Vec<Card> = Vec::with_capacity(total);
    for (i, mut c) in sessions.into_iter().enumerate() {
        c.days_idle = days_since(&c.last_activity);
        c.stale = c.days_idle.map(|d| d > 30).unwrap_or(false);
        let s = summarize(&c, use_llm).await;
        c.status = s.status;
        c.left_off = s.left_off;
        c.next_step = s.next_step;
        c.blockers = s.blockers;
        cards.push(c);
        if (i + 1) % 10 == 0 || i + 1 == total {
            eprintln!("[harvest] summarized {}/{}", i + 1, total);
        }
    }

    // trello business cards (already classified)
    for mut t in trello::collect() {
        t.days_idle = days_since(&t.last_activity);
        t.stale = t.days_idle.map(|d| d > 30).unwrap_or(false);
        cards.push(t);
    }

    let (new_i, changed_i) = store::upsert_cards(&cards)?;
    eprintln!("[harvest] db: {} new issues, {} changed", new_i, changed_i);

    let board = store::export_board()?;
    write_board(&board)?;
    eprintln!(
        "[harvest] wrote tasklord.db + board.json — {} cards {:?}",
        board.total, board.counts
    );
    Ok(())
}

fn write_board(board: &crate::model::Board) -> Result<()> {
    let dir = config::board_dir();
    fs::create_dir_all(&dir)?;
    let payload = serde_json::to_string_pretty(board)?;
    fs::write(dir.join("board.json"), &payload)?;
    fs::write(dir.join("board.js"), format!("window.BOARD = {};\n", payload))?;
    Ok(())
}
